lib/core/services/sort-by-category.service.ts

```diff
@@ -0,0 +1,104 @@
+/*!
+ * @license
+ * Copyright 2019 Alfresco Software, Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Injectable } from '@angular/core';
+
+export interface SortbleByCategoryItem {
+  name: string;
+  category?: string;
+}
+
+export interface ItemsByCategory<T> {
+  category: string;
+  items: T[];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SortByCategoryMapperService<T extends SortbleByCategoryItem = SortbleByCategoryItem> {
+
+  private defaultCategories: string[] = [];
+
+  mapItems(items: T[], defaultCategories: string[]): ItemsByCategory<T>[] {
+    this.defaultCategories = defaultCategories;
+
+    const sortedItems = this.sortItems(items);
+    const itemsByCategory = this.mapItemsByCategory(sortedItems);
+    const itemsSortedByCategory = this.sortCategories(itemsByCategory);
+
+    return itemsSortedByCategory;
+  }
+
+  private mapItemsByCategory(items: T[]): ItemsByCategory<T>[] {
+    const itemsByCategoryObject: { [category: string]: T[] } = {};
+
+    items.forEach((item) => {
+      const category = this.mapItemsCategoryForDisplaying(item);
```

Comment:
Thinking if we can change name of the `mapItemsCategoryForDisplaying` to e.g. `mapItemDefaultCategory`

The service do not know if we are going to display these data or not

Reply:
Changed